detailing the use of meta-information for activities tracking and dev… (#1426)
OWASP/owasp-mstg: Document/0x04e-Testing-Authentication-and-Session-Management.md

### Testing Login Activity and Device Blocking (MSTG-AUTH-11)

For applications which require L2 protection, the MASVS states that they should inform the user about all login activities within the app with the possiblity of blocking certain devices. This can be broken down into various scenarios:

> 1. The application provides a push notification the moment their account is used on another device to notify the user of different activities. The user can then block this device after opening the app via the push-notification. 

Reviewer: No trailing spaces: can you please run the linter there?
```suggestion
1. The application provides a push notification the moment their account is used on another device to notify the user of different activities. The user can then block this device after opening the app via the push-notification.
```

2. The application provides an overview of the last session after login, if the previous session was with a different configuration (e.g. location, device, app-version) then the user his current configuration. The user then has the option to report suspicious activities and block devices used in the previous session.
3. The application provides an overview of the last session after login at all times.
4. The application has a self-service portal in which the user can see an audit-log and manage the different devices with which he can login.

The developer can make use of specific meta-information and associate it to each different activity or event within the application. This will make it easier for the user to spot suspicious behavior and block the corresponding device. The meta-information may include:

- Device: The user can clearly identify all devices where the app is being used.
- Date and Time: The user can clearly see the latest date and time when the app was used.
- Location: The user can clearly identify the lastest locations where the app was used.

> The application can provide a push notification after each sensitive activity within the application. The choice of which activities to audit needs to be done for each application based on the data it handles and the level of security risk the team is willing to have. Below is a list of common sensitive activies that are usually audited:

Reviewer: This could also just be in the list of activities/account history as an alternative to push notifications :). (e.g. see more than just the last location, but more or less a list of locations used: it should be a balance between privacy and security depending on the risks involved)

- Login attempts
- Password changes
- PII changes (name, email address, telephone number, etc.)

> - Sensitive activities (purchase, accessing important ressources, etc.)

Reviewer: as part of #1382, can you add `consent to Terms & Conditions` here?

Reviewer: thanks!

> Paid content requires special care, and additional meta-information (e.g., operation Cost, credit, etc.) might be used to ensure user's knowledge about the whole operation's parameters.   

Reviewer: Not sure what you mean by this?

Author: This part is like an additional note as an emphasis about logging activities for paid content with additional meta-information to ensure that paid content are well protected

> 

Reviewer: LAst, maybe this is the place, maybe not: but part of #1382 is to note that the actions (e.g. consent, buying stuff, etc. should be signed so you can see that these sensitive actions actually happened by that user)

3. The application provides an overview of the last session after login at all times.
4. The application has a self-service portal in which the user can see an audit-log and manage the different devices with which he can login.

The developer can make use of specific meta-information and associate it to each different activity or event within the application. This will make it easier for the user to spot suspicious behavior and block the corresponding device. The meta-information may include:

- Device: The user can clearly identify all devices where the app is being used.
- Date and Time: The user can clearly see the latest date and time when the app was used.
- Location: The user can clearly identify the lastest locations where the app was used.

The application can provide a list of activities history which will be updated after each sensitive activity within the application. The choice of which activities to audit needs to be done for each application based on the data it handles and the level of security risk the team is willing to have. Below is a list of common sensitive activies that are usually audited:

- Login attempts
- Password changes

> - PII changes (name, email address, telephone number, etc.)

Reviewer 2: Style guide: as PII has only one occurrence in the file we should write it in full.

```suggestion
- Personal Identifiable Information changes (name, email address, telephone number, etc.)
```

- Sensitive activities (purchase, accessing important ressources, etc.)

> - Consent to Terms & Conditions

Reviewer 2: ```suggestion
- Consent to Terms and Conditions clauses
```

Paid content requires special care, and additional meta-information (e.g., operation Cost, credit, etc.) might be used to ensure user's knowledge about the whole operation's parameters.

> In addition and to ensure non-repudiation, actions (E.g. payed content access, given consent to T&Cs, etc.) should be signed when necessary to be sure that these actions are being performed by the exact user.

Reviewer 2: In non-repudiation, ideally the user performing a transaction will be authenticated and the integrity of the transaction is protected. I'd rather leave the used mechanisms open. If not, we'll have to go into more detail. What do you think about the following change?

```suggestion
In addition, non-repudiation mechanisms should be applied to sensitive transactions (e.g. payed content access, given consent to Terms and Conditions clauses, etc.) in order to prove that a specific transaction was in fact performed (integrity) and by whom (authentication).```

### Testing Login Activity and Device Blocking (MSTG-AUTH-11)

> For applications which require L2 protection, the MASVS states that: "The app informs the user of all login activities with their account. Users are able view a list of devices used to access the account, and to block or allow specific devices.". 

Reviewer 2: I would not include here what the MASVS says, if we start putting this in several test cases we end up having to take care of updating these snippets as well :/

Author: Makes sense. i will remove this and remplace it with something similar but not from the MASVS.

> 

Reviewer: removing these 4 items might not be a good idea.

Author: the thing was that the items were re-explained diffrently and i tried to add more detailed. but we can keep them i just wanted to avoid redundancy .

Reviewer 2: I'd go for a merge of both things, as both have good content :)
